frontend/apps/filemanager/filemanager.lua

```diff
@@ -470,6 +479,35 @@ function FileManager:onSwipeFM(ges)
     return true
 end
 
+function FileManager:addFileDialogButtons(row_id, row_func) -- FileManager, History, Collections file_dialog
+    self.file_dialog_added_buttons = self.file_dialog_added_buttons or { index = {} }
+    if self.file_dialog_added_buttons.index[row_id] == nil then
+        table.insert(self.file_dialog_added_buttons, row_func)
+        self.file_dialog_added_buttons.index[row_id] = #self.file_dialog_added_buttons
+    end
+end
+
+function FileManager:removeFileDialogButtons(row_id)
+    local index = self.file_dialog_added_buttons and self.file_dialog_added_buttons.index[row_id]
+    if index ~= nil then
+        table.remove(self.file_dialog_added_buttons, index)
+        if #self.file_dialog_added_buttons == 0 then
+            self.file_dialog_added_buttons = nil
+        else
+            self.file_dialog_added_buttons.index[row_id] = nil
+            for id, idx in pairs(self.file_dialog_added_buttons.index) do
```

Comment:
(I thing using pairs() is right - just asking so you too are sure we should not use ipairs().)

Reply:
`index` is a hashmap table, { row_id = _row index in the buttons table_ }.